### PR TITLE
Sanitize repo and namespace names

### PR DESCRIPTION
**Pull request by newswangerd**
Fixes: #429 
Fixes: #692 
Fixes: #662

Namespaces:
- Convert existing namespaces with `-` in the name to `_`
- Ensure names don't start with `_` and are more than 2 characters
- Lowercase namespaces for new github users

Repos:
- Lowercase name and replace `-` and `.` with `_` during new repo creation
- Ensure names don't start with `_` and are more than 2 characters

Content:
- For single content repositories, set content name to repo.original name
- Added repository to unique fields

### CODE REVIEW

File: galaxy/main/signals/handlers.py [hunk 1]
```diff
@@ -41,15 +41,17 @@ def user_logged_in_handler(request, user, **kwargs):
...
 
-    if user.namespaces.count():
-        # User is associated with one or more namespaces.
-        # Even though the Namespaces may be inactive, we'll assume
-        # the user has been here before, and knows how to manage Namespaces.
+    username = social.extra_data.get('login')
+    sanitized_username = username.lower().replace('-', '_')
+
+    try:
+        namespace = models.ProviderNamespace.objects.get(name=username)
```
Comment (chouseknecht) [inline galaxy/main/signals/handlers.py]: Think the filter should be name__iexact=username, to account for existing namespaces that may have contain uppercase letters.

File: galaxy/main/signals/handlers.py [hunk 2]
```diff
@@ -59,7 +61,7 @@ def user_logged_in_handler(request, user, **kwargs):
 
         if provider.name.lower() == 'github':
```
Comment (chouseknecht) [inline galaxy/main/signals/handlers.py]: I think we no longer need to iterate over Providers. We assumed GitHub above to get the social object (line #39), so might as well continue to assume GitHub here as well, and use the existing social object.

File: galaxy/worker/importers/base.py [hunk 3]
```diff
@@ -50,8 +50,10 @@ def make_content(self):
         repo = self.ctx.repository
         ns = repo.provider_namespace.namespace
 
-        name = repo.name
-        original_name = repo.name
+        # Name is the content name, which in the case of multi content repos
+        # might not be the same as the repository name
+        name = repo.original_name
```
Comment (chouseknecht) [inline galaxy/worker/importers/base.py]: Name should default to the repo.name, which should have been sanitized of any '-' and '.' chars and converted to lowercase() somewhere.

File: galaxy/main/migrations/0108_sanitize_namespace_names.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.13 on 2018-05-31 13:18
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+STRIP_VERSION_SUFFIX = """
```
Comment (chouseknecht) [inline galaxy/main/migrations/0108_sanitize_namespace_names.py]: Nit picky, I know, but this variable name should be related to the task at hand. Look like it was copy-pasta'd from some other migration.

File: galaxy/api/views/namespace.py [hunk 5]
```diff
@@ -48,9 +48,13 @@
 def check_basic(data, errors):
     if not data.get('name'):
         errors['name'] = "Attribute 'name' is required"
-    elif not re.match('^[\w-]+$', data['name']):
-        # Allow only names containing word chars and '-'
-        errors['name'] = "Name can only contain [A-Za-z0-9-_]"
+    elif not re.match('^[\w]+$', data.get('name')):
+        # Allow only names containing word chars
+        errors['name'] = "Name can only contain [A-Za-z0-9_]"
+    elif(len(data.get('name')) <= 2):
```
Comment (cutwater) [inline galaxy/api/views/namespace.py]: Possible `TypeError` here. You're using `data.get('name')` multiple times.
Use `name = data.get('name')` in the beginning of the function and work with a variable below.

File: galaxy/api/views/namespace.py [hunk 6]
```diff
@@ -48,9 +48,13 @@
...
+        errors['name'] = "Name must be longer than 2 characters"
+    elif(data.get('name')[0] == '_'):
```
Comment (cutwater) [inline galaxy/api/views/namespace.py]: `name.startswith()`

File: galaxy/api/views/repository.py [hunk 7]
```diff
@@ -62,13 +62,17 @@ def get_repo(provider_namespace, user, repo_name):
 def check_name(name):
     if not name:
         raise ValidationError(detail={'name': 'Name is required'})
+    elif not re.match('^[\.\w-]+$', name):
```
Comment (cutwater) [inline galaxy/api/views/repository.py]: No need for `elif`, raise will interrupt function execution.

File: galaxy/main/migrations/0108_sanitize_namespace_names.py (hunk 4 above)
Comment (cutwater) [reply galaxy/main/migrations/0108_sanitize_namespace_names.py]: +1

**Pull request merged**